profile/src/wallet/wallet.rs

```diff
@@ -0,0 +1,372 @@
+        wallet.wallet_client_storage.save(
+            SecureStorageKey::DeviceFactorSourceMnemonic {
+                factor_source_id: private_hd_factor_source
+                    .factor_source
+                    .id
+                    .clone(),
+            },
+            &private_hd_factor_source.mnemonic_with_passphrase,
+        )?;
+        Ok(wallet)
+    }
+
+    /// Creates wallet by *importing* a Profile.
+    #[uniffi::constructor]
+    pub fn by_importing_profile(
+        profile: Profile,
+        secure_storage: Arc<dyn SecureStorage>,
+    ) -> Self {
+        Wallet::init_logging();
+
+        log::info!(
+            "Instantiating Wallet by importing a Profile with ID: {}",
+            profile.id()
+        );
+
+        Self::with_imported_profile(profile, secure_storage)
+    }
+
+    #[uniffi::constructor]
+    pub fn by_loading_profile(
+        secure_storage: Arc<dyn SecureStorage>,
+    ) -> Result<Self> {
+        Wallet::init_logging();
+
+        log::info!(
+            "Instantiating Wallet by loading the active Profile from storage"
+        );
+
+        // Init WalletClient's storage
+        let wallet_client_storage = WalletClientStorage::new(secure_storage);
+
+        // Load active profile ID
+        let active_profile_id: ProfileID = wallet_client_storage.load_or(
+            SecureStorageKey::ActiveProfileID,
+            CommonError::NoActiveProfileIDSet,
+        )?;
+
+        Self::new_load_profile_with_id(active_profile_id, wallet_client_storage)
+    }
+
+    #[uniffi::constructor]
+    pub fn by_loading_profile_with_id(
+        profile_id: ProfileID,
+        secure_storage: Arc<dyn SecureStorage>,
+    ) -> Result<Self> {
+        Wallet::init_logging();
+
+        log::info!(
+            "Instantiating Wallet by loading the Profile with ID {} from storage",
+            profile_id
+        );
+
+        Self::new_load_profile_with_id(
+            profile_id,
+            WalletClientStorage::new(secure_storage),
+        )
+    }
+}
+
+#[cfg(test)]
+impl Wallet {
+    pub(crate) fn ephemeral(
+        profile: Profile,
+    ) -> (Self, Arc<EphemeralSecureStorage>) {
+        let storage = EphemeralSecureStorage::new();
+        (
+            Self::by_importing_profile(profile, storage.clone()),
+            storage,
+        )
+    }
+}
+#[cfg(test)]
+impl HasPlaceholder for Wallet {
+    fn placeholder() -> Self {
+        Self::ephemeral(Profile::placeholder()).0
+    }
+
+    fn placeholder_other() -> Self {
+        Self::ephemeral(Profile::placeholder_other()).0
+    }
+}
+
+//========
+// GET
+//========
+#[uniffi::export]
+impl Wallet {
+    /// Takes a snapshot of the profile and serialize it as a String of JSON.
+    pub fn json_snapshot(&self) -> String {
+        serde_json::to_string(&self.profile())
+            .expect("Should always be able to JSON serialize a Profile.")
+    }
+
+    /// Clone the profile and return it.
+    pub fn profile(&self) -> Profile {
```

Comment:
See comment above about `json_snapshot()` only used for DEBUG features like view JSON of Profile... 

Actually we might even remove this method, clients might never have to read out Profile as a whole?